pkg/route/ingress/ingress.go

```diff
@@ -583,13 +583,25 @@ func targetPortForService(namespace string, path *extensionsv1beta1.HTTPIngressP
 		expect := path.Backend.ServicePort.StrVal
 		for _, port := range service.Spec.Ports {
 			if port.Name == expect {
+				if port.TargetPort.Type == intstr.String {
+					targetPort := intstr.FromString(port.Name)
+					return &targetPort
+				}
 				return &port.TargetPort
 			}
 		}
 	} else {
 		for _, port := range service.Spec.Ports {
 			expect := path.Backend.ServicePort.IntVal
 			if port.Port == expect {
+				if port.TargetPort.Type == intstr.String {
+					if &port.Name == nil {
+						// TODO lookup numeric port in endpoint
```

Comment:
@Miciah maybe you can help here - this is the special case where the Service port is unnamed but references the Pod port via name. Here the controller has to lookup the target port in the corresponding Endpoint.

Reply:
A service port must have a name if the service has more than 1 port.  If the service does not have more than 1 port, then the route's `spec.to.name` field can be left unspecified, and the route will use whatever port exists on the endpoints resource.

If a service port does have a name, then the corresponding port on the endpoints resource will have the same name, so I think it would be reasonable always to use the service port name if it exists, and otherwise omit `spec.to.name` from the route per the above reasoning.  Does that make sense?

I find this tricky to reason about, so in case it is helpful to you or others, I'll elaborate on my reasoning.  Each service port either has a name or does not, and the port specifies its target either by name or by number.  Thus we have four cases to consider:

| Service port `name` | Service port `targetPort` | Route `spec.to.name`     |
|------------------- |------------------------- |------------------------ |
| absent              | number                    | empty                    |
| absent              | name                      | empty                    |
| present             | number                    | port name or target port |
| present             | name                      | port name                |

In the first two cases, we do not have a service port name, but the service can have only 1 port (because if it had multiple ports, the ports would need to have names), so specifying a value for the route's `spec.to.name` field would be superfluous anyway.

In the third case, the service may have multiple ports (thus we need to set the route's `spec.to.name` field), we have a service port name and a target port number, and either one could be used to specify the route's `spec.to.name` field. I suggest using the service port name for consistency with the fourth case (below), but I could accept the argument that it would be preferable to use the target port number for consistency with the current logic.

In the fourth case, the service may have multiple ports (same as the third case), and we have a port name but no target port number, so we must use the port name for the route's `spec.to.name`. (Note that this is the case that is broken with the current code, because the current code uses the target port name, which may differ from the port name on the endpoints resource.)